openvdb/openvdb/tools/FastSweeping.h

```diff
@@ -1406,7 +1631,19 @@ struct FastSweeping<SdfGridT, ExtValueT>::SweepingKernel
                     if (update <= d2.v) {
                         if (update < absV) {
                             value = sign * update;
-                            if (acc2) acc2->setValue(ijk, acc2->getValue(d1(ijk)));//update ext?
+                            if (acc2) {
+                                // There is an assert upstream to check if mExtGridInput exists if mode != SWEEP_DEFAULT
+                                ExtValueT updateExt = acc2->getValue(d1(ijk));
+                                if (mode == FastSweepingDirection::SWEEP_GREATER_THAN_ISOVALUE) {
```

Comment:
I'm concerned about these two extra branch tests in the inner-most loop of the old Fast Sweeping algorithm. Have you tested the performance impact? Is there no way of extracting those if-statements into three different kernels - at least so the 99% default cause is not negatively impacted by the new (arguably rare) cases?

Reply:
I had a test with extending a Cd grid from a sphere to a box with 215 million voxels. Before this change, the timing is 12.613s. After the change, I get the following timing:

- SWEEP_ALL_DIRECTION = 12.645s
- SWEEP_GREATER_THAN_ISOVALUE = 12.929s
- SWEEP_LESS_THAN_ISOVALUE = 12.652s.

So sweeping in a given direction does make the algorithm go slower by about 2% in this case. 

I think we can rid of these two extra branches by making mode as a template and making the SOP to dispatch the correct instantiated template based on the user input. Let me know what you think. 

![EXTRAPOLATE_TEST1](https://user-images.githubusercontent.com/4682534/133300277-5904982d-0798-48d8-a160-9215399e700d.png)

![EXTRAPOLATE_TEST1_OPENGL](https://user-images.githubusercontent.com/4682534/133300297-33b24ce3-a42b-456a-87b7-ff65d31e621c.png)

![EXTRAPOLATE_TEST1_BENCHMARK](https://user-images.githubusercontent.com/4682534/133300325-69d5fb1e-58e3-450a-aa2a-a1cd9bfda46e.png)

Reply:
@apradhana  thanks for looking into this - I think 2% is small enough that it shouldn't block this PR (I'll approve it).  At some later time it would be great if you could allow for the the two code path to be selected at compile-time (say with a boolean template parameter) so as to avoid the (granted) small overhead.

Reply:
Thank you @kmuseth. I started taking a stab at it, but couldn't finish it on time. That's my homework for the next iteration.